streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -836,10 +836,27 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                      final OptimizableRepartitionNodeBuilder<K1, V1> optimizableRepartitionNodeBuilder) {
 
 
-        final String repartitionTopic = repartitionTopicNamePrefix + REPARTITION_TOPIC_SUFFIX;
-        final String sinkName = builder.newProcessorName(SINK_NAME);
-        final String nullKeyFilterProcessorName = builder.newProcessorName(FILTER_NAME);
-        final String sourceName = builder.newProcessorName(SOURCE_NAME);
+        final String repartitionTopicName = repartitionTopicNamePrefix.endsWith(REPARTITION_TOPIC_SUFFIX) ?
+            repartitionTopicNamePrefix : repartitionTopicNamePrefix + REPARTITION_TOPIC_SUFFIX;
+
+        // Always need to generate the names to burn index counter for compatibility
+        final String genSinkName = builder.newProcessorName(SINK_NAME);
+        final String genNullKeyFilterProcessorName = builder.newProcessorName(FILTER_NAME);
+        final String genSourceName = builder.newProcessorName(SOURCE_NAME);
+
+        final String sinkName;
+        final String sourceName;
+        final String nullKeyFilterProcessorName;
+
+        if (repartitionTopicNamePrefix.matches("KSTREAM.*-[0-9]{10}")) {
```

Comment:
It seems like we could know whether the user supplied the topic name or not and pass that as a parameter to this function, in which case we would not need to do any pattern matching?

Reply:
I had thought of that approach when doing this PR, but I didn't want to have the logic spread out in different areas of the code.  If you insist I can make the change to pass in a `boolean` flag indicating it's a user-provided name.

Reply:
My intuition is that it would be more direct and less brittle to pass the flag. I suspect you would agree? There is a cost associated with the change as you pointed. It looks like this would impact a couple of call paths and `OptimizableRepartitionNode`? If so, that doesn't look too bad. I don't have a good intuition as to whether `KSTREAM-xxxxxxxxxx` is reserved and how likely we are to change that.

Given the information I have right now, I would make the change. I welcome you or someone else with more experience than me (a low bar :)) to assert that I've under estimated the cost or that the code is so unlikely to change that the benefit does not outweigh the cost. If that's the case I'm willing to accept it as is.

Reply:
Fair enough, I'll go ahead and make the changes to pass in a flag indicating a user has provided the name.